pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+            """
+            Set the value of the ``Context`` object this socket is tied to.
+            This operation (changing the context) may not always be supported.
+            """
+
+        @abstractproperty
+        def negotiated_tls_version(self) -> Optional[TLSVersion]:
+            """
+            The version of TLS that has been negotiated on this connection.
+            """
+
+
+Cipher Suites
+~~~~~~~~~~~~~
+
+Todo
+
+Protocol Negotiation
+~~~~~~~~~~~~~~~~~~~~
+
+Both NPN and ALPN allow for protocol negotiation as part of the HTTP/2
+handshake. While NPN and ALPN are, at their fundamental level, built on top of
+bytestrings, string-based APIs are frequently problematic as they allow for
+errors in typing that can be hard to detect.
+
+For this reason, this module would define a type that protocol negotiation
+implementations can pass and be passed. This type would wrap a bytestring, but
+allow for aliases for well-known protocols. This allows us to avoid the
+problems inherent in typos for well-known protocols, while allowing the full
+extensibility of the protocol negotiation layer if needed.
+
+::
+
+    NextProtocol = namedtuple('NextProtocol', ['name'])
+
+    H2 = NextProtocol(b'h2')
+    H2C = NextProtocol(b'h2c')
+    HTTP1 = NextProtocol(b'http/1.1')
+    WEBRTC = NextProtocol(b'webrtc')
+    C_WEBRTC = NextProtocol(b'c-webrtc')
+    FTP = NextProtocol(b'ftp')
+    STUN = NextProtocol(b'stun.nat-discovery')
+    TURN = NextProtocol(b'stun.turn')
+
+TLS Versions
+~~~~~~~~~~~~
+
+It is often useful to be able to restrict the versions of TLS you're willing to
+support. There are many security advantages in refusing to use old versions of
+TLS, and some misbehaving servers will mishandle TLS clients advertising
+support for newer versions.
+
+The following enumerated type can be used to gate TLS versions. Forward-looking
+applications should almost never set a maximum TLS version unless they
+absolutely must, as a TLS backend that is newer than the Python that uses it
+may support TLS versions that are not in this enumerated type.
+
+Additionally, this enumerated type defines two additional flags that can always
+be used to request either the lowest or highest TLS version supported by an
+implementation.
+
+::
+
+    class TLSVersion(Enum):
+        MINIMUM_SUPPORTED
+        SSLv2
+        SSLv3
+        TLSv1
+        TLSv1_1
+        TLSv1_2
+        TLSv1_3
+        MAXIMUM_SUPPORTED
+
+
+Errors
+~~~~~~
+
+This module would define three base classes for use with error handling. Unlike
+the other classes defined here, these classes are not *abstract*, as they have
+no behaviour. They exist simply to signal certain common behaviours. Backends
+should subclass these exceptions in their own packages, but needn't define any
+behaviour for them. These exceptions should *never* be thrown directly, they
+should always be subclassed.
+
+The definitions of the errors are below::
+
+    class TLSError(Exception):
```

Comment:
Yup, it is. But I was worried about getting too specific here and constraining individual implementations. In general I'd suggest we only want to break out errors for things that applications could meaningfully recover from: otherwise they're all equally fatal.